### PR TITLE
Make use of EntityGraph in katharsis-jpa optional #380

### PR DESCRIPTION
**Pull request by remmeier**
moved code in question into a dedicated class that is only accessed with reflection when actually necessary.

### CODE REVIEW

File: katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java [hunk 1]
```diff
@@ -175,30 +174,26 @@ public boolean equals(Object obj) {
 	}
 
 	protected void applyFetchPaths(Query criteriaQuery) {
-		EntityGraph<T> graph = em.createEntityGraph(getEntityClass());
-		for (MetaAttributePath fetchPath : fetchPaths) {
-			applyFetchPaths(graph, fetchPath);
+		if (!fetchPaths.isEmpty()) {
+			EntityGraphBuilder builder;
+			try {
+				// avoid compile-time dependency
```
Comment (chb0github) [inline katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java]: Yeah - I see that. Why?

File: katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java [hunk 2]
```diff
@@ -95,8 +95,7 @@ public MetaDataObject getMeta() {
 				entityList.add((T) values[0]);
```
Comment (chb0github) [inline katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java]: this whole function `getResultList()` can be reduced to 
`return isCompoundSelection() ? list.stream().map(Object[].class::cast).map(o -> o[0]).collect(toList()) : (List<T>)list;`

File: katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java (hunk 1 above)
Comment (remmeier) [inline katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java]: added a longer comment to reflect what is written in the ticket

File: katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java (hunk 2 above)
Comment (remmeier) [inline katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java]: source compatiblity is still Java 7, but built with Java 8. Various people asked for it the past few weeks. I think should be ok this way, community growth seems more important than the latest Java 8 features. But we could start using Java 8 for test cases and possibily for non-core parts like new modules.

as disclaimer: I don't like this, but it seems to be the pace of a significant fraction of the java enterprise community. Just this week I saw Katharsis 3 master running perfectly fine in a six year old app server.

Comment (chb0github) [inline katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java]: java 8 isn't _new_ - java 9 is due out this summer. This isn't about performance. It's more about reduction of the code base.

Comment (remmeier) [inline katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java]: welcome to the enterprise community...

Lambdas and stuff are nice, but there are currently also plenty of other things to worry.

File: katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java [hunk 3]
```diff
@@ -175,30 +174,30 @@ public boolean equals(Object obj) {
...
-			applyFetchPaths(graph, fetchPath);
+		if (!fetchPaths.isEmpty()) {
+			EntityGraphBuilder builder;
+			try {
+				// avoid compile-time dependency since EntityGraph is not used by default
+				// only when requested in API (by adding fetchPaths). Some more advanced
+				// customizations of JPA module likely will make use of it, but basic features
+				// work without it, meaning that katharsis-jpa also works in older JPA implementations
+				// not supporting the fairly new EntityGraph
+				builder = (EntityGraphBuilder) Class.forName(ENTITY_GRAPH_BUILDER_IMPL).newInstance();
```
Comment (hibaymj) [inline katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractQueryExecutorImpl.java]: Why are you invoking reflections on every call through here?

**Pull request closed without merging**